### ISSUE TITLE
uint8 scalar tensors cannot be used for integer indexing

### ISSUE THREAD

**Issue by pmeier** [pytorch/pytorch#70916]
Integer, scalar tensors should behave like integers when used as index. Tensors of dtype `torch.uint8` deviate from that:

```python
import torch

t_1d_single = torch.empty(1)
t_1d_multi = torch.empty(2)

for dtype in (torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64):
    print("single", dtype, t_1d_single[torch.tensor(0, dtype=dtype)].shape)
    print("multi1", dtype, t_1d_multi[torch.tensor(0, dtype=dtype)].shape)
    print("multi2", dtype, t_1d_multi[torch.tensor(1, dtype=dtype)].shape)
    print("#" * 50)
```

```
single torch.uint8 torch.Size([0, 1])
multi1 torch.uint8 torch.Size([0, 2])
multi2 torch.uint8 torch.Size([1, 2])
##################################################
single torch.int8 torch.Size([])
multi1 torch.int8 torch.Size([])
multi2 torch.int8 torch.Size([])
##################################################
single torch.int16 torch.Size([])
multi1 torch.int16 torch.Size([])
multi2 torch.int16 torch.Size([])
##################################################
single torch.int32 torch.Size([])
multi1 torch.int32 torch.Size([])
multi2 torch.int32 torch.Size([])
##################################################
single torch.int64 torch.Size([])
multi1 torch.int64 torch.Size([])
multi2 torch.int64 torch.Size([])
##################################################
```

**Comment by albanD** [comment]
I think this is related to the fact that `uint8` used to be our type used for masking before bool was introduced.
But masking with `uint8` is deprecated, you should have a warning right? So maybe it is time to forbid it completely. So that in next release we can enable indexing with it?

**Comment by pmeier** [comment]
> you should have a warning right?

Nope.

**Comment by mruberry** [comment]
We would accept a PR that improves the deprecation warnings for uint8 indexing.

**Comment by ngimel** [comment]
For a special case of 0d indexing tensor, we don't warn and silently convert it to long index before sending to `index` function https://github.com/pytorch/pytorch/blob/9950f3b7e645803f3f25a6f53c1b88e376ff2c1d/aten/src/ATen/TensorIndexing.h#L414. It's not particularly efficient, and it fails to trigger a warning that other uint8 indexing invocations trigger.
